README
NAME

        sixtemp_i2c.h - client library for sixtemp sensor

SYNOPSIS

        #include <Arduino.h>
        #include <sixtemp_i2c.h>
        #include <Wire.h>
    
        sixtemp_i2c stemp(0x18);
    
        void setup () {
            Serial.begin(9600);
            stemp.begin();
        }
    
        void loop () {
            stemp.refresh();
            Serial.println("sensor count: "+String(stemp.count));
            if (stemp.count) {
                sixtemp_i2c_temp temp = stemp.get_temp(0);
                Serial.println("temp: "+temp.address_as_string()+" "+temp.as_string());
            }
    
            delay(3000);
        }

DESCRIPTION

    This is a client library for sixtemp sensor
    https://github.com/jozef/sixtemp_sensor

INSTALL

    Clone the repository folder into sketchbook/libraries/sixtemp_i2c

CLASS

 class sixtemp_i2c

  sixtemp_i2c(uint8_t i2c_addr = DEFAULT_I2C_ADDR);

    Class constructor.

  void begin();

    Start i2c.

  sixtemp_i2c_temp get_temp(uint8_t idx);

    Return "class sixtemp_i2c_temp" for temperature at idx.

  void refresh();

    Refresh all temperature readings from sixtemp sensor via i2c.

  uint8_t count;

    Number of temperatures present.

  sixtemp_i2c_temp temp_sensors[MAX_SENSORS];

    Array of all "class sixtemp_i2c_temp".

 class sixtemp_i2c_temp

    Represents one temperature. Has following public variables and methods:

  bool has_error;

    true/false if there was an error

  bool has_address;

    true/false if there temp position has address assigned

  float temp_c;

    temperature

  String as_string();

    temp_c as String, ex.: "21.8 C"

  char address[8];

    Temperature sensors 8 byte address.

  String address_as_string();

    address as String, ex.: "28ff3ff6a1160322"

EXAMPLES

 examples/sixtemp_readout/sixtemp_readout.ino

    Example serial output when uploaded to an Arduino connected to
    sixtemp_sensor via i2c:

        sensor count: 6
        sensor 1: n/a
        sensor 2: 28ff3ff6a1160322 19.8 C
        sensor 3: 286164122e7c499a 21.8 C
        sensor 4: 286164122f88f163 error
        sensor 5: 28ffa59492160558 19.8 C
        sensor 6: 286164122f89682a 21.5 C
        readouts count: 92, errors: 9

SEE ALSO

    https://github.com/jozef/sixtemp_sensor

LICENSE

    This is free software, licensed under the MIT License.

AUTHOR

        Jozef Kutej



Changelog
0.01 2018-01-12
    - initial version


sixtemp_i2c.cpp
/* see sixtemp_i2c.h */

#include <sixtemp_i2c.h>
#include <Arduino.h>
#include <Wire.h>

#define TRANSMISSION_REQUEST_DELAY 100

String sixtemp_i2c_temp::as_string() {
    if (has_error) {
        return "error";
    }

    if (!has_address) {
        return "n/a";
    }

    return String(int(temp_c))
        +'.'
        +String(int((temp_c - int(temp_c))*10))
        +" C";
}

String sixtemp_i2c_temp::address_as_string() {
    if (!has_address) {
        return "n/a";
    }

    char addr_as_char[2*sizeof(address)+1];
    addr_as_char[2*sizeof(address)+1-1] = '\0';
    for (uint8_t i = 0; i < 16; i++) {
        uint8_t digit = address[i >> 1];
        digit = (i & 1 ? (digit & 0x0F) : (digit >> 4));
        addr_as_char[i] = (digit > 9 ? ('a'+digit-10) : ('0'+digit));
    }

    return String(addr_as_char);
}

sixtemp_i2c::sixtemp_i2c(uint8_t i2c_addr) {
    _i2c_addr = i2c_addr;
    count = 0;
    memset(temp_sensors,0,sizeof(temp_sensors));
}

void sixtemp_i2c::begin() {
    Wire.begin();
    refresh();
}

sixtemp_i2c_temp sixtemp_i2c::get_temp(uint8_t idx) {
    return temp_sensors[idx];
}

void sixtemp_i2c::set_led(bool flag) {
    Wire.beginTransmission(_i2c_addr);
    Wire.write(0x21);
    Wire.write(flag);
    Wire.endTransmission();
    delayMicroseconds(TRANSMISSION_REQUEST_DELAY);
}

bool sixtemp_i2c::led_status() {
    Wire.beginTransmission(_i2c_addr);
    Wire.write(0x01);
    Wire.endTransmission();
    delayMicroseconds(TRANSMISSION_REQUEST_DELAY);
    Wire.requestFrom(_i2c_addr, (uint8_t)1);
    return Wire.read();
}

void sixtemp_i2c::refresh() {
    Wire.beginTransmission(_i2c_addr);
    Wire.write(0x60); // read sensors count
    Wire.endTransmission();
    delayMicroseconds(TRANSMISSION_REQUEST_DELAY);
    Wire.requestFrom(_i2c_addr, (uint8_t)1);
    count = Wire.read();
    if (count > 0x80) {
        count = 0;
        return;
    }
    if (count > MAX_SENSORS) {
        count = MAX_SENSORS;
    }
    for (uint8_t i = 0; i < count; i++) {
        Wire.beginTransmission(_i2c_addr);
        Wire.write(0x60+1+i); // set sensor idx
        Wire.endTransmission();
        delayMicroseconds(TRANSMISSION_REQUEST_DELAY);
        Wire.requestFrom(_i2c_addr, sizeof(temp_sensors[0]));
        for (uint8_t i2 = 0; i2 < sizeof(temp_sensors[0]); i2++) {
            if (!Wire.available()) {
                temp_sensors[i].has_error = true;
                temp_sensors[i].has_address = false;
                break;
            }
            char rch = Wire.read();
            *((char*)&temp_sensors[i] + i2) = rch;
        }
    }
    for (uint8_t i = count; i < MAX_SENSORS; i++) {
        if (temp_sensors[i].has_address) {
            temp_sensors[i].has_error = true;
        }
    }
}


sixtemp_i2c.h
/* VERSION 0.01; 1.1.2018; see below for description and documentation */
#ifndef sixtemp_i2c_h
#define sixtemp_i2c_h

#include <Arduino.h>

#define MAX_SENSORS 6
#define DEFAULT_I2C_ADDR 0x18

class sixtemp_i2c_temp {
    public:
        bool has_error;
        bool has_address;
        float temp_c;
        char address[8];
        String as_string();
        String address_as_string();
};

class sixtemp_i2c {
    public:
        sixtemp_i2c(uint8_t i2c_addr = DEFAULT_I2C_ADDR);
        void begin();
        sixtemp_i2c_temp get_temp(uint8_t idx);
        void refresh();
        uint8_t count;
        sixtemp_i2c_temp temp_sensors[MAX_SENSORS];
        void set_led(bool flag);
        bool led_status();
    private:
        uint8_t _i2c_addr;
};

#endif

/*

=head1 NAME

    sixtemp_i2c.h - client library for sixtemp sensor

=head1 SYNOPSIS

    #include <Arduino.h>
    #include <sixtemp_i2c.h>
    #include <Wire.h>

    sixtemp_i2c stemp(0x18);

    void setup () {
        Serial.begin(9600);
        stemp.begin();
    }

    void loop () {
        stemp.refresh();
        Serial.println("sensor count: "+String(stemp.count));
        if (stemp.count) {
            sixtemp_i2c_temp temp = stemp.get_temp(0);
            Serial.println("temp: "+temp.address_as_string()+" "+temp.as_string());
        }

        delay(3000);
    }

=head1 DESCRIPTION

This is a client library for sixtemp sensor L<https://github.com/jozef/sixtemp_sensor>

=head1 INSTALL

Clone the repository folder into F<sketchbook/libraries/sixtemp_i2c>

=head1 CLASS

=head2 class sixtemp_i2c

=head3 sixtemp_i2c(uint8_t i2c_addr = DEFAULT_I2C_ADDR);

Class constructor.

=head3 void begin();

Start i2c.

=head3 sixtemp_i2c_temp get_temp(uint8_t idx);

Return L</class sixtemp_i2c_temp> for temperature at C<idx>.

=head3 void refresh();

Refresh all temperature readings from sixtemp sensor via i2c.

=head3 uint8_t count;

Number of temperatures present.

=head3 sixtemp_i2c_temp temp_sensors[MAX_SENSORS];

Array of all L</class sixtemp_i2c_temp>.

=head2 class sixtemp_i2c_temp

Represents one temperature. Has following public variables and methods:

=head3 bool has_error;

true/false if there was an error

=head3 bool has_address;

true/false if there temp position has address assigned

=head3 float temp_c;

temperature

=head3 String as_string();

temp_c as String, ex.: "21.8 C"

=head3 char address[8];

Temperature sensors 8 byte address.

=head3 String address_as_string();

address as String, ex.: "28ff3ff6a1160322"

=head1 EXAMPLES

=head2 examples/sixtemp_readout/sixtemp_readout.ino

Example serial output when uploaded to an Arduino connected to sixtemp_sensor
via i2c:

    sensor count: 6
    sensor 1: n/a
    sensor 2: 28ff3ff6a1160322 19.8 C
    sensor 3: 286164122e7c499a 21.8 C
    sensor 4: 286164122f88f163 error
    sensor 5: 28ffa59492160558 19.8 C
    sensor 6: 286164122f89682a 21.5 C
    readouts count: 92, errors: 9

=head1 SEE ALSO

L<https://github.com/jozef/sixtemp_sensor>

=head1 LICENSE

This is free software, licensed under the MIT License.

=head1 AUTHOR

    Jozef Kutej

=cut

*/
